src/firmware/main/math/polynomial_2d.c

```diff
@@ -0,0 +1,178 @@
+                                                                                           \
+        Polynomial2dOrder##N_minus_1##_t deriv =                                           \
+            shared_polynomial2d_differentiateOrder##N(p);                                  \
+                                                                                           \
+        /* We divide dt by two here because we want to ensure we get the number of */      \
+        /* values that the user requested, but simpsons rule requires that we use */       \
+        /* an even number of points */                                                     \
+        float dt = (t_max - t_min) / ((parametrization.num_values - 1) * 2);               \
+                                                                                           \
+        /* Populate the entries of the parametrization by numerically integrating the */   \
+        /* derivative with Simpson's rule: */                                              \
+        /* https://www.intmath.com/integration/6-simpsons-rule.php */                      \
+        float initial_value        = calculateL2NormAtValueOrder##N_minus_1(deriv, t_min); \
+        float sum_of_odd_s_values  = 0;                                                    \
+        float sum_of_even_s_values = 0;                                                    \
+        parametrization.s_values[0] = 0;                                                   \
+        parametrization.t_values[0] = t_min;                                               \
+        for (size_t i = 1; i < parametrization.num_values; i++)                            \
+        {                                                                                  \
+            const float t_i     = t_min + 2 * i * dt;                                      \
+            const float value_i = calculateL2NormAtValueOrder##N_minus_1(deriv, t_i);      \
+                                                                                           \
+            const float t_i_minus_1 = t_min + (2 * i - 1) * dt;                            \
+            const float value_i_minus_1 =                                                  \
+                calculateL2NormAtValueOrder##N_minus_1(deriv, t_i_minus_1);                \
+            sum_of_odd_s_values += value_i_minus_1;                                        \
+                                                                                           \
+            const float s_i = initial_value + 4 * sum_of_odd_s_values +                    \
+                              2 * sum_of_even_s_values + value_i;                          \
+                                                                                           \
+            parametrization.s_values[i] = s_i;                                             \
```

Comment:
I might be missing something, but since this function takes ` ArcLengthParametrization_t parametrization` as an argument and not `ArcLengthParametrization_t* parametrization`, changing these values won't get propagated to the caller like we would want them to?

Reply:
Heh, but an array is just a pointer in C. So copying the struct just means you copy the pointer to the array, not the array itself.

More seriously, I was a bit concerned this would be confusing. Changed to take a pointer instead.

Reply:
Actually thinking a bit more, this does make things a bit ickier in the test just visually, what do you think?

Reply:
ah right I see it now, adding the dereference operator everywhere will be an eye sore, probably can leave it as is